### PR TITLE
[server] sequentially gc workspaces

### PR DESCRIPTION
**Pull request by svenefftinge**
## Description

Makes the workspace-gc job garbage collect workspaces sequentially.

<details>
<summary>Summary generated by Copilot</summary>

### <samp>🤖 Generated by Copilot at 8e2f8a1</samp>

Improved workspace garbage collection logic and logging. Replaced `Promise.all` with `for` loops for better performance and error handling. Added more logging and error handling to `WorkspaceGarbageCollector` in `components/server/src/jobs/workspace-gc.ts`.

</details>

## Related Issue(s)

Fixes EXP-389

#### Preview status

gitpod:summary

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

File: components/server/src/jobs/workspace-gc.ts [hunk 1]
```diff
@@ -80,16 +90,21 @@ export class WorkspaceGarbageCollector implements Job {
...
             const afterSelect = new Date();
-            const deletes = await Promise.all(
-                workspaces.map((ws) => this.workspaceService.deleteWorkspace(SYSTEM_USER, ws.id, "gc")),
-            );
+            log.info(`workspace-gc: about to soft-delete ${workspaces.length} workspaces`);
+            for (const ws of workspaces) {
+                try {
+                    await this.workspaceService.deleteWorkspace(SYSTEM_USER, ws.id, "gc");
+                } catch (err) {
+                    log.error({ workspaceId: ws.id }, "workspace-gc: error during workspace soft-deletion", err);
```
Comment (akosyakov) [inline components/server/src/jobs/workspace-gc.ts:98]: Does it make sense to count failures and subtract from total?

Comment (svenefftinge) [reply components/server/src/jobs/workspace-gc.ts:98]: no, I don't think so. we will see the error logs in that case

**Review by akosyakov — APPROVED** [review]
code changes look good, but it did not try

**Comment by AlexTugarev** [comment]
Let's try with the preview env first.

**Comment by svenefftinge** [comment]
let's try in prod, as it it is not user-facing

/unhold

**Comment by svenefftinge** [comment]
> Let's try with the preview env first.

oh, sorry didn't see that message. but I believe it would b quite hard to retry as you need to create workspaces that are softdeleted and also over retention period. Also we'd want to have that on a large load.

**Comment by svenefftinge** [comment]
but also I'm surprised it got merged although the build didn't yet run through 👀

**Pull request merged**